### PR TITLE
feat: enable NRF response caching

### PR DESCRIPTION
**Pull request by gatici**
Implementing NRF response caching together with NF subscribe and notify services

### CODE REVIEW

**Review by gab-arrobo — COMMENTED** [review]
Hi @gatici, I started reviewing this PR and it looks to me that it will increase the complexity of managing the dependencies because, with this PR, we would have the following dependency "chain": config5g/openapi -> nrf -> pcf. So, updates in the `pcf` would depend on new tags for the `nrf` (I would like to point out that something similar is already happening in the `amf` repo). Just in case, have you explored the possibility of moving the `nrf/nrfcache` code (this only depends on the `openapi` repo) into the `openapi` repo such that we can avoid the dependency chain?

File: callback/api_nf_subscribe_notify.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// SPDX-FileCopyrightText: 2022 Infosys Limited
+// Copyright 2019 free5GC.org
```
[the remark this answers is not in the document]
Comment (thakurajayL) [reply callback/api_nf_subscribe_notify.go:2]: Infosys folks had developed the nrf cache at start,

[the remark this answers is not in the document]
Comment (gatici) [reply callback/api_nf_subscribe_notify.go:2]: Hello @gab-arrobo, @thakurajayL ,
1.  NRF caching code is moved to Openapi library through the PR: https://github.com/omec-project/openapi/pull/9/files. Please review. 
2.  The code is copied from AMF as it already implemented there. Does it make sense to update the header as below  or should we keep the first contributors ? I need you advice.

// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2024 Canonical Ltd.
/*

**Comment by gab-arrobo** [comment]
@gatici, please rebase this PR based on the changes done in the `openapi` repo. You can use the tag that was recently created: https://github.com/omec-project/openapi/releases. Thanks!

**Comment by gatici** [comment]
> @gatici, please rebase this PR based on the changes done in the `openapi` repo. You can use the tag that was recently created: https://github.com/omec-project/openapi/releases. Thanks!

Hello Gabriel,

Thank you for updating the tag.  PR is updated by using the `openapi` repo and `nrf` repository is removed from the dependencies. Please review. Thanks.

**Review by gab-arrobo — COMMENTED** [review]
Hi @gatici, have you tested the changes? I am asking this because I just finished to run some tests using `Aether-OnRamp` (having `enableNrfCaching` set to `true` and then set to `false`) and in both scenarios, the tests failed. All profiles failed due to "registration reject" (`procedure:REGISTRATION-PROCEDURE, error:triggering event:REGISTRATION-REQUEST-EVENT, expected event:AUTHENTICATION-REQUEST-EVENT, received event:REGISTRATION-REJECT-EVENT"`)

```
time="2024-08-08T18:31:05Z" level=info msg="Profile Name: profile1 , Profile Type: register"
time="2024-08-08T18:31:05Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T18:31:05Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T18:31:07Z" level=info msg="Profile Name: profile2 , Profile Type: pdusessest"
time="2024-08-08T18:31:07Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T18:31:07Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T18:31:09Z" level=info msg="Profile Name: profile3 , Profile Type: anrelease"
time="2024-08-08T18:31:09Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T18:31:09Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T18:31:15Z" level=info msg="Profile Name: profile4 , Profile Type: uetriggservicereq"
time="2024-08-08T18:31:15Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T18:31:15Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T18:31:22Z" level=info msg="Profile Name: profile5 , Profile Type: deregister"
time="2024-08-08T18:31:22Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T18:31:22Z" level=info msg="Profile Status: FAIL"
time="2024-08-08T18:31:23Z" level=info msg="Profile Name: profile7 , Profile Type: uereqpdusessrelease"
time="2024-08-08T18:31:23Z" level=info msg="Ue's Passed: 0 , Ue's Failed: 5"
time="2024-08-08T18:31:23Z" level=info msg="Profile Status: FAIL"
```

**Comment by gatici** [comment]
Hello @gab-arrobo 
The issue is fixed and PR is validated by E2E tests. I have made some refactoring to add unit tests. There are some an oversight which causes the problem. Now this is solved.
```
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [TRAC][GNBSIM][SimUe][imsi-208930100007487] Sending Procedure Result to Profile : PASS
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [TRAC][GNBSIM][SimUe][imsi-208930100007487] Sending PROC-PASS-EVENT to Profile routine
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [TRAC][GNBSIM][SimUe][imsi-208930100007487] Sent  PROC-PASS-EVENT to Profile routine
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][ProfUeCtx][imsi-208930100007487] imsiStateMachine received result
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][ProfUeCtx][imsi-208930100007487] procedure: USER-DATA-PACKET-GENERATION-PROCEDURE, status: PROC-PASS-EVENT, E2E latency [ms]: 3004
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][ProfUeCtx][imsi-208930100007487] Procedure Result: PASS, imsi: imsi-208930100007487
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][Profile][profile1] No more procedures left
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][ProfUeCtx][imsi-208930100007487] imsiStateMachine no more proc to execute
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][ProfUeCtx][imsi-208930100007487] imsiStateMachine ended
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][Profile][profile1] ExecuteProfile ended
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][Summary] Profile Name: profile1 , Profile Type: pdusessest
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][Summary] Ue's Passed: 1 , Ue's Failed: 0
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:37Z [INFO][GNBSIM][Summary] Profile Status: **PASS**
2024-08-15T01:22:47.466Z [container-agent] 2024-08-15T01:22:42Z [INFO][GNBSIM][Stats] Dump all metrics
```

File: pcftests/pcfcfg.yaml [hunk 2]
```diff
@@ -4,6 +4,7 @@
 configuration:
   defaultBdtRefId: BdtPolicyId-
   nrfUri: https://nrf:443
+  enableNrfCaching: true
```
Comment (gab-arrobo) [inline pcftests/pcfcfg.yaml:7]: Should we add the variable in the yaml file and set a default value? Should it be `900 (15mins)` as it done in the `init_context.go`?
```suggestion
  enableNrfCaching: true
  nrfCacheEvictionInterval: <some-value>
```

File: producer/bdtpolicy.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func getBDTPolicyContextProcedure(bdtPolicyID string) (
 	} else {
 		// not found
 		problemDetail := util.GetProblemDetail("Can't find bdtPolicyID related resource", util.CONTEXT_NOT_FOUND)
-		logger.Bdtpolicylog.Warnf(problemDetail.Detail)
+		logger.Bdtpolicylog.Warnln(problemDetail.Detail)
```
Comment (gab-arrobo) [inline producer/bdtpolicy.go:61]: I am curious, why this change?

File: producer/callback.go [hunk 4]
```diff
@@ -39,6 +48,59 @@ func HandleSmPolicyNotify(request *httpwrapper.Request) *httpwrapper.Response {
 	return httpwrapper.NewResponse(http.StatusNotImplemented, nil, nil)
 }
 
-// TODO: handle SM Policy Notify
+// SmPolicyNotifyProcedure TODO: handle SM Policy Notify
 func SmPolicyNotifyProcedure(supi string, notification models.PolicyDataChangeNotification) {
 }
+
+func HandleNfSubscriptionStatusNotify(request *httpwrapper.Request) *httpwrapper.Response {
```
Comment (gab-arrobo) [inline producer/callback.go:55]: Should we add a comment to the function for documentation purposes?

File: producer/callback.go [hunk 5]
```diff
@@ -39,6 +48,59 @@ func HandleSmPolicyNotify(request *httpwrapper.Request) *httpwrapper.Response {
...
+
+	problemDetails := NfSubscriptionStatusNotifyProcedure(notificationData)
+	if problemDetails != nil {
+		return httpwrapper.NewResponse(int(problemDetails.Status), nil, problemDetails)
+	} else {
+		return httpwrapper.NewResponse(http.StatusNoContent, nil, nil)
+	}
+}
+
+func NfSubscriptionStatusNotifyProcedure(notificationData models.NotificationData) *models.ProblemDetails {
```
Comment (gab-arrobo) [inline producer/callback.go:68]: Should we add a comment to the function for documentation purposes?

File: producer/policyauthorization.go [hunk 6]
```diff
@@ -108,7 +108,7 @@ func handleMediaSubComponent(smPolicy *pcf_context.UeSmPolicyData, medComp *mode
 				var ul, dl bool
 				qosData, ul, dl = updateQosInMedSubComp(smPolicy.PolicyDecision.QosDecs[qosID], medComp, medSubComp)
 				if problemDetails := modifyRemainBitRate(smPolicy, &qosData, ul, dl); problemDetails != nil {
-					logger.PolicyAuthorizationlog.Errorf(problemDetails.Detail)
+					logger.PolicyAuthorizationlog.Errorln(problemDetails.Detail)
```
Comment (gab-arrobo) [inline producer/policyauthorization.go:111]: Same as in the other comment, I am curious to know why this change?

File: Makefile [hunk 7]
```diff
@@ -58,7 +58,7 @@ $(GO_BIN_PATH)/%: %.go $(NF_GO_FILES)
 # $(@F): The file-within-directory part of the file name of the target.
 	@echo "Start building $(@F)...."
 	cd $(GO_SRC_PATH)/ && \
-	CGO_ENABLED=0 go build -ldflags "$(LDFLAGS)" -o $(ROOT_PATH)/$@ $(@F).go
+	CGO_ENABLED=0 go build -gcflags='all=-N -l' -ldflags "$(LDFLAGS)" -o $(ROOT_PATH)/$@ $(@F).go
```
[the remark this answers is not in the document]
Comment (gatici) [reply Makefile:61]: gcflags are removed.

File: callback/api_nf_subscribe_notify.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+package callback
+
+import (
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/omec-project/openapi/models"
+	"github.com/omec-project/pcf/logger"
+
```
[the remark this answers is not in the document]
Comment (gatici) [reply callback/api_nf_subscribe_notify.go:16]: Empty line is removed and imports are sorted.

File: callback/api_nf_subscribe_notify.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/omec-project/openapi"
+	"github.com/omec-project/pcf/producer"
```
[the remark this answers is not in the document]
Comment (gatici) [reply callback/api_nf_subscribe_notify.go:18]: They are sorted.

File: callback/api_nf_subscribe_notify.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/omec-project/util/httpwrapper"
+)
+
+func HTTPNfSubscriptionStatusNotify(c *gin.Context) {
+	var nfSubscriptionStatusNotification models.NotificationData
+
+	requestBody, err := c.GetRawData()
+	if err != nil {
+		logger.CallbackLog.Errorf("Get Request Body error: %+v", err)
```
[the remark this answers is not in the document]
Comment (gatici) [reply callback/api_nf_subscribe_notify.go:27]: Done

File: producer/bdtpolicy.go (hunk 3 above)
Comment (gatici) [reply producer/bdtpolicy.go:61]: In fails when we run the linter. As there is not non-constant string, we need to use  Warnln. If we use f-string, the we need to Warnf.

producer/bdtpolicy.go:61:29: printf: non-constant format string in call to (*github.com/sirupsen/logrus.Entry).Warnf (govet)
                logger.Bdtpolicylog.Warnf(problemDetail.Detail)

File: producer/policyauthorization.go (hunk 6 above)
Comment (gatici) [reply producer/policyauthorization.go:111]: This fails in govet and aims to fix linting issues.

File: pcftests/pcfcfg.yaml (hunk 2 above)
Comment (gatici) [reply pcftests/pcfcfg.yaml:7]: Default value is already set in the init_context.go, hence we did not add  a variable here. However, we could do it.

Comment (gatici) [reply pcftests/pcfcfg.yaml:7]: Done

File: consumer/nf_management.go [hunk 11]
```diff
@@ -154,3 +155,63 @@ var SendUpdateNFInstance = func(patchItem []models.PatchItem) (nfProfile models.
...
+	client := Nnrf_NFManagement.NewAPIClient(configuration)
+	var res *http.Response
+
+	res, err = client.SubscriptionIDDocumentApi.RemoveSubscription(context.Background(), subscriptionId)
+	if err == nil {
+		return
+	} else if res != nil {
+		defer func() {
+			if bodyCloseErr := res.Body.Close(); bodyCloseErr != nil {
+				err = fmt.Errorf("RemoveSubscription' response body cannot close: %w", bodyCloseErr)
```
[the remark this answers is not in the document]
Comment (gatici) [reply consumer/nf_management.go:205]: Done

File: consumer/nf_discovery.go [hunk 12]
```diff
@@ -14,41 +14,72 @@ import (
...
-	nrfUri string, targetNfType, requestNfType models.NfType, param Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (
-	*models.SearchResult, error,
+var (
+	CreateSubscription        = SendCreateSubscription
+	NRFCacheSearchNFInstances = nrfCache.SearchNFInstances
+	StoreApiSearchNFInstances = (*Nnrf_NFDiscovery.NFInstancesStoreApiService).SearchNFInstances
+)
+
+var SendSearchNFInstances = func(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (
+	models.SearchResult, error,
```
[the remark this answers is not in the document]
Comment (gatici) [reply consumer/nf_discovery.go:30]: https://github.com/omec-project/openapi/blob/main/nrfcache/nrfcache.go#L348 
In openapi nrfcache library  `func SearchNFInstances` has a different method signature than the one exists in this file. We need to use this method from common repository, so this arrangement is required.  AMF is also have some signature which is done in this PR. That's reason we required to change some parameter types.

SearchNFInstances from nrfcache library:
```
func SearchNFInstances(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (models.SearchResult, error) 
```

File: factory/config.go [hunk 13]
```diff
@@ -49,7 +49,9 @@ type Configuration struct {
 	DnnList   map[string][]string // sst+sd os key
 	SlicePlmn map[string]PlmnSupportItem
 
-	PlmnList []PlmnSupportItem `yaml:"plmnList,omitempty"`
+	PlmnList                 []PlmnSupportItem `yaml:"plmnList,omitempty"`
+	EnableNrfCaching         bool              `yaml:"enableNrfCaching"`
+	NrfCacheEvictionInterval int               `yaml:"nrfCacheEvictionInterval,omitempty"`
```
[the remark this answers is not in the document]
Comment (gatici) [reply factory/config.go:54]: We think 900 second is reasonable.

File: pcftests/pcf_test.go [hunk 14]
```diff
@@ -221,9 +234,9 @@ func TestRegisterNF(t *testing.T) {
 		fmt.Printf("Test RegisterNFInstance called\n")
 		return prof, "", "", nil
 	}
-	consumer.SendSearchNFInstances = func(nrfUri string, targetNfType, requestNfType models.NfType, param Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (*models.SearchResult, error) {
+	consumer.SendSearchNFInstances = func(nrfUri string, targetNfType, requestNfType models.NfType, param *Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (models.SearchResult, error) {
```
[the remark this answers is not in the document]
Comment (gatici) [reply pcftests/pcf_test.go:237]: https://github.com/omec-project/openapi/blob/main/nrfcache/nrfcache.go#L348
In openapi nrfcache library func SearchNFInstances has a different method signature than the one exists in this file. We need to use this method from common repository, so this parameter types arrangement is required.  AMF SendSearchNFInstances func also  has the  same signature which is done in this PR. 

As explained above, after changing the parameter types of SendSearchNFInstances func in nf_discovery.go file, same arrangement is required in the tests.

File: producer/callback.go (hunk 5 above)
Comment (gatici) [reply producer/callback.go:68]: Comment  is added.

File: producer/callback.go (hunk 4 above)
Comment (gatici) [reply producer/callback.go:55]: Comment is added.

File: consumer/nf_discovery.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (gab-arrobo) [reply consumer/nf_discovery.go:30]: Understood. Thanks!

**Review by gab-arrobo — APPROVED** [review]
+1 (tested changes with OnRamp having `enableNrfCaching` enabled and then `enableNrfCaching` disabled)
